### PR TITLE
COM-20002: Bundles Design improvements

### PR DESCRIPTION
**Pull request by kmadejski**
**JIRA issues**: 
1. [https://jira.ez.no/browse/COM-20002](https://jira.ez.no/browse/COM-20002)
2. [https://jira.ez.no/browse/COM-20010](https://jira.ez.no/browse/COM-20010)
3. [https://jira.ez.no/browse/COM-20011](https://jira.ez.no/browse/COM-20011)

### Description
This PR contains design changes ordered in mentioned JIRA issue.

### Bundle FieldType configuration:
| Field name | Field ID | Type | Required | Searchable | Translatable |
|-------------|--------|------|----------|------------|--------------|
| Bundle ID | bundle_id | ezstring | Yes | Yes | No |
| Name | name | ezstring | No | Yes | Yes |
| Description | description | ezrichtext | Yes | Yes | Yes |
| Packagist URL | packagist_url | ezurl | Yes | No | No |
| Downloads | downloads | ezinteger | No | Yes | No |
| Stars | stars | ezinteger | No | Yes | No |
| Forks | forks | ezinteger |No | Yes | No |
| Updated | updated | ezdate | No | Yes | No |

### Installation notes
1. Make sure that we have *Solr* enabled and properly configured on production.
2. Change name of page (ContentType: BundlesList): from **Bundles** to **eZ Platform Bundles**.
3. Check if ContentTypes configuration is compatible with the list above.
4. Update production installation.
5. Run: `composer install`
6. Run: `php app/console app:update_bundles_list`
7. (Should be done automatically during `composer install`): Run: `php app/console ezplatform:reindex`
8. Clear cache.
9. Clear your browser cache.

### CODE REVIEW

File: app/Resources/views/full/bundle_list.html.twig [hunk 1]
```diff
@@ -6,14 +6,57 @@
...
+                        <li class="active"><a href="#">All Bundles</a></li>
+                        <li><a href="#">Content Type</a></li>
+                        <li><a href="#">Field Type</a></li>
+                        <li><a href="#">Theme</a></li>
+                        <li><a href="#">Social</a></li>
+                        <li><a href="#">e-Commerce</a></li>
+                        <li><a href="#">Administration</a></li>
+                        <li><a href="#">Workflow</a></li>
+                    </ul>
+                    <h2>Compatibility</h2>
```
Comment (SylvainGuittard) [inline app/Resources/views/full/bundle_list.html.twig]: You can remove the `Compatibility` section. The first stable for v2 is planned for fall 2017. We might implement this filter in 2018.

File: app/Resources/views/line/bundle_card.html.twig [hunk 2]
```diff
@@ -1,47 +1,66 @@
...
-                <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}" title="See on Packagist.org">{{ ez_render_field(content, 'bundle_id') }}</a>
-            </h2>
-            <div class="bundle-card-line-description">
-                {{ ez_render_field(content, 'description') }}
+<div class="bundle-card-line-container{% if extra.page is defined %} page-{{ extra.page }}{% endif %}">
+    <div class="bundle-card-line">
+        <div class="bundle-card-line-data">
+            <div class="bundle-card-line-information">
+                <h1>
+                    <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}">
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: I missed this in the previous PR.
Why not using `packagist_url` field?

Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Good observation! I will fix it.

File: app/Resources/views/line/bundle_card.html.twig [hunk 3]
```diff
@@ -1,47 +1,66 @@
...
+                            <div class="bundle-card-line-maintainer-username">{{ maintainer.name }}</div>
                         </div>
                     {% endfor %}
-            </div>
-            {% if packageDetails.maintainers|length > 4 %}
-                <div class="bundle-card-line-contributors-more">
-                    <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}" title="See on Packagist.org">+{{ packageDetails.maintainers|length - 4 }} more</a>
+                    {% if packageDetails.maintainers|length > 3 %}
+                        <div class="bundle-card-line-contributors-more">
+                            <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}" title="See on Packagist.org">+{{ packageDetails.maintainers|length - 3 }} more</a>
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: same as https://github.com/ezsystems/ezplatform-com/pull/101/files#r123730866

File: app/Resources/views/line/bundle_card.html.twig [hunk 4]
```diff
@@ -1,47 +1,66 @@
...
+                        </div>
+                    {% endif %}
                 </div>
-            {% endif %}
+                {% endif %}
+            </div>
+        </div>
+        <div class="bundle-card-line-href">
+            <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}" title="See on Packagist.org"><i class="fa fa-angle-right fa-3" aria-hidden="true"></i></a>
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: same as https://github.com/ezsystems/ezplatform-com/pull/101/files#r123730866

File: app/Resources/views/line/bundle_card.html.twig [hunk 5]
```diff
@@ -1,47 +1,64 @@
...
-                <a href="https://packagist.org/packages/{{ ez_field_value(content, 'bundle_id') }}" title="See on Packagist.org">{{ ez_render_field(content, 'bundle_id') }}</a>
-            </h2>
-            <div class="bundle-card-line-description">
-                {{ ez_render_field(content, 'description') }}
+<div class="bundle-card-line-container{% if extra.page is defined %} page-{{ extra.page }}{% endif %}">
+    <div class="bundle-card-line">
+        <div class="bundle-card-line-data">
+            <div class="bundle-card-line-information">
+                <h1>
+                    {{ ez_render_field(content, 'packagist_url') }}
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: Finally, I think `h1` should be clickable.
```
<a href="{{ ez_content_name (content) }}"> {{ ez_render_field(content, 'packagist_url') }} </a>
```

Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: Hmm... It is. Please notice that I'm using `ez_render_field` here and it generates URL automatically.

File: app/config/services/twig.yml [hunk 6]
```diff
@@ -5,3 +5,7 @@ services:
             - '@ezpublish.api.service.content_type'
         tags:
             - name: twig.extension
+    twig.extension.text:
```
Comment (SylvainGuittard) [inline app/config/services/twig.yml]: I don't see where this is used.

File: app/Resources/views/line/bundle_card.html.twig (hunk 5 above)
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: oh! good point :)

File: app/config/services/twig.yml (hunk 6 above)
Comment (kmadejski) [inline app/config/services/twig.yml]: This is required for Twig `truncate` function which is used here: [https://github.com/ezsystems/ezplatform-com/pull/101/files#diff-205461f0c399ee7cdb4ed7a7eda49f76R36](https://github.com/ezsystems/ezplatform-com/pull/101/files#diff-205461f0c399ee7cdb4ed7a7eda49f76R36)

File: app/Resources/views/full/bundle_list.html.twig [hunk 7]
```diff
@@ -6,14 +6,56 @@
...
+                            </div>
+                        </div>
+                    {{ form_end(searchForm) }}
+                </div>
+            </div>
+        </div>
         <div class="bundles-list-content">
             <div class="container">
-                <div class="col-xs-12 bundles-list-container">
+                <div class="hidden-xs hidden-sm col-md-2 bundle-list-settings">
```
Comment (SylvainGuittard) [inline app/Resources/views/full/bundle_list.html.twig]: I think this part can be removed. This feature will be implemented on a different PR

File: app/Resources/views/full/bundle_list.html.twig [hunk 8]
```diff
@@ -6,14 +6,56 @@
...
+                        <li><a href="#">Social</a></li>
+                        <li><a href="#">e-Commerce</a></li>
+                        <li><a href="#">Administration</a></li>
+                        <li><a href="#">Workflow</a></li>
+                    </ul>
+                </div>
+                <div class="col-xs-12 col-md-10 bundles-list-container">
+                    <div class="row">
+                        <div class="col-xs-12 col-md-9">
+                            <h2>All Bundles</h2>
```
Comment (SylvainGuittard) [inline app/Resources/views/full/bundle_list.html.twig]: I know that this part was not specified in the `Story`, I forgot the result page.
I think this heading should be "connected" to the search: after submitting the form, it's not `All Bundles`.
Is it possible to replace `All Bundles` by `Search / Keyword user` when the user searched for `user`?

File: app/Resources/views/line/bundle_card.html.twig [hunk 9]
```diff
@@ -1,47 +1,62 @@
...
-                        <img src="{{ packageDetails.authorAvatarUrl }}?s=100" alt="{{ packageDetails.author[0].name }}" />
-                    </div>
-                {% endif %}
-                {% if packageDetails.updated is defined %}
-                    <div class="bundle-card-line-last-update">Updated: <span>{{ packageDetails.updated|date("d / m / Y") }}</span></div>
-                {% endif %}
+            <div class="bundle-card-line-counters">
+                <ul>
+                    <li class="downloads"><i class="fa fa-cloud-download" aria-hidden="true"></i> {{ ez_field_value(content, 'downloads') }}</li>
+                    <li class="stars"><i class="fa fa-star" aria-hidden="true"></i> {{ ez_field_value(content, 'stars') }}</li>
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: In the jira issue (https://jira.ez.no/browse/COM-20010), I associated `stars` to `popularity`. Does it make sense?
@damianz5 @kmadejski What do you think?

File: app/Resources/views/full/bundle_list.html.twig [hunk 10]
```diff
@@ -6,14 +6,56 @@
...
+                        </div>
+                        <div class="col-xs-12 col-md-3 sort-container">
+                            {{ form_start(sortingForm) }}
+                                <div class="form-group {% if not sortingForm.order.vars.valid %}has-error{% endif %}">
+                                    {{  form_row(sortingForm.order) }}
```
Comment (SylvainGuittard) [inline app/Resources/views/full/bundle_list.html.twig]: In the demo this morning, I saw `stars`.
In the jira issue (https://jira.ez.no/browse/COM-20010), I associated `stars` to `popularity`. Does it make sense?
@damianz5 @kmadejski What do you think? Difficult to change?

File: app/Resources/views/line/bundle_card.html.twig [hunk 11]
```diff
@@ -1,47 +1,62 @@
...
+                    <li class="forks"><i class="fa fa-code-fork" aria-hidden="true"></i> {{ ez_field_value(content, 'forks') }}</li>
+                    <li class="updated"><i class="fa fa-clock-o" aria-hidden="true"></i> {{ ez_field_value(content, 'updated')|date("d / m / Y") }}</li>
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: I think we can save some space using 
`|date("d/m/Y")`

File: app/Resources/views/line/bundle_card.html.twig [hunk 12]
```diff
@@ -1,47 +1,62 @@
...
-    {% if packageDetails.maintainers is defined %}
-        <div class="hidden-xs col-sm-12 bundle-card-line-contributors">
-            <div class="bundle-card-line-contributors-container">
-                <span>By:</span>
-                    {% for maintainer in packageDetails.maintainers|slice(0, 4) %}
-                        <div class="bundle-card-line-contributor">
-                            <div class="bundle-card-line-contributor-photo">
+            <div class="bundle-card-line-authors">
+                <div class="bundle-card-line-owner">
+                    {% if packageDetails.authorAvatarUrl is defined %}
```
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: Why not having the profile picture and the name of the Owner under the same `if` condition?
If something happens and you don't have the picture, the page will display only the name and not `<span>Owner:</span>`

File: app/config/routing.yml [hunk 13]
```diff
@@ -36,17 +36,35 @@ _multiFileUpload:
     prefix:   "%ezpublish_rest.path_prefix%"
 
 _ezplatform_blog_ajax:
-    path: /ezplatform/blog/load/page/{page}
+    path: /ezplatform/blog/load/page/{page}/{order}
     defaults: { _controller: app.controller.blog:showBlogPostsAction }
     methods:  [GET]
     requirements:
         "page": "\d+"
 _ezplatform_bundles_list_ajax:
```
Comment (SylvainGuittard) [inline app/config/routing.yml]: CS? empty lines between route definitions?

File: composer.json [hunk 14]
```diff
@@ -44,7 +44,7 @@
         "ezsystems/content-on-the-fly-prototype": "~0.1.7",
         "willdurand/js-translation-bundle": "^2.6.4",
         "ezsystems/comments-bundle": "~6.0",
-        "knplabs/packagist-api": "~1.3",
+        "knplabs/packagist-api": "dev-master",
```
Comment (SylvainGuittard) [inline composer.json]: Because they did not tag since your PR?

File: src/AppBundle/Command/UpdateBundlesListCommand.php [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $repository = $this->getContainer()->get('ezpublish.api.repository');
+        $packagistServiceProvider = $this->getContainer()->get('app.packagist_service_provider');
+        $userService = $repository->getUserService();
+        $searchService = $repository->getSearchService();
+        $contentService = $repository->getContentService();
+        $permissionResolver = $repository->getPermissionResolver();
+
+        $user = $userService->loadUserByLogin('admin');
```
Comment (SylvainGuittard) [inline src/AppBundle/Command/UpdateBundlesListCommand.php]: Not a big fan of `admin`... 
Do we have another option?
Can we use something like `$repository->sudo`?

File: src/AppBundle/Command/UpdateBundlesListCommand.php [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+                $contentUpdateStruct->setField('description', $xmlText);
+
+                $contentDraft = $contentService->updateContent($contentDraft->versionInfo, $contentUpdateStruct);
+                $contentService->publishVersion($contentDraft->versionInfo);
+
+                $output->writeln(': Updated');
+            }
+            else {
+                $output->writeln(': OK');
```
Comment (SylvainGuittard) [inline src/AppBundle/Command/UpdateBundlesListCommand.php]: We can maybe improve this message ;)
What about `: Already up-to-date`?

File: src/AppBundle/Command/UpdateBundlesListCommand.php [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        }
+        $output->writeln("All bundles has been successfully updated.");
```
Comment (SylvainGuittard) [inline src/AppBundle/Command/UpdateBundlesListCommand.php]: It's not all the bundles ;)
typo: `$output->writeln("The bundles have been successfully updated.");`

File: src/AppBundle/Form/OrderType.php [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+     * @param array $options
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('order', ChoiceType::class, array(
+                'placeholder' => 'Sort by',
+                'choices' => array(
+                    'latestUpdate' => 'Latest Update',
+                    'stars' => 'Stars',
```
Comment (SylvainGuittard) [inline src/AppBundle/Form/OrderType.php]: See: https://github.com/ezsystems/ezplatform-com/pull/101/files#r125325618

File: src/AppBundle/Service/Packagist/PackagistServiceProvider.php [hunk 19]
```diff
@@ -81,14 +85,18 @@ private function callApi($packageName)
...
+        $packageDetails['stars'] = $externalData->getGithubStars();
         $versions = $externalData->getVersions();
         if (is_array($versions) && !empty($versions)) {
             if (isset($versions['dev-master'])) {
                 $version = 'dev-master';
             } else {
                 $version = key($versions);
             }
-            $packageDetails['updated'] = $versions[$version]->getTime();
+            $packageDetails['updated'] = \DateTime::createFromFormat(\DateTime::ISO8601, $versions[$version]->getTime());
```
Comment (SylvainGuittard) [inline src/AppBundle/Service/Packagist/PackagistServiceProvider.php]: Not sure to understand this line. We don't use timestamp?

File: web/assets/css/bundle.css [hunk 20]
```diff
@@ -1,171 +1,258 @@
...
+.bundle-card-line-maintainer-photo {
+    width: 24px;
+    height: 24px;
+    float: left;
 }
 
-.bundle-card-line-contributors span {
-    display: block;
+.bundle-card-line-owner-photo {
+    margin: 0 10px 0 10px;
```
Comment (SylvainGuittard) [inline web/assets/css/bundle.css]: Improvement: can be `margin: 0 10px;`

File: web/assets/css/bundle.css [hunk 21]
```diff
@@ -1,171 +1,258 @@
...
+    .bundle-card-line-owner-name {
+        display: none;
+    }
+
+    .bundle-card-line-maintainers {
+        padding-right: 0px;
+    }
+
+    .bundle-card-line-maintainer {
+        margin: 0 5px 0 5px;
```
Comment (SylvainGuittard) [inline web/assets/css/bundle.css]: improvement: `margin: 0 5px;`

File: web/assets/css/bundle.css [hunk 22]
```diff
@@ -1,171 +1,258 @@
...
+
+    .bundle-card-line-owner {
+        width: 100%;
+    }
+
+    .bundle-card-line-maintainers {
+        width: 100%;
+        padding: 0;
+    }
+}
```
Comment (SylvainGuittard) [inline web/assets/css/bundle.css]: CS: please add an empty line ;)

File: app/Resources/views/full/bundle_list.html.twig [hunk 23]
```diff
@@ -6,14 +6,56 @@
             <div class="container">
                 <div class="row">
                     <div class="col-xs-12">
-                        {{ ez_render_field(content, 'description') }}
+                        <h1>{{ ez_render_field(content, 'name') }}</h1>
+                        <h2>{{ ez_render_field(content, 'description') }}</h2>
```
Comment (SylvainGuittard) [inline app/Resources/views/full/bundle_list.html.twig]: This cannot be `<h2>` because it's a RichText.

File: app/Resources/views/line/bundle_card.html.twig (hunk 5 above)
Comment (SylvainGuittard) [inline app/Resources/views/line/bundle_card.html.twig]: It should not be `h1` because a web page "should" only have one `h1`. So for all the bundle cards, please change to `<h2>`

File: web/assets/css/bundles_list.css [hunk 24]
```diff
@@ -1,18 +1,153 @@
...
+}
+
+.bundles-list-search input#bundles-list-search-query {
+    height: 50px;
+    font-size: 21px;
+    font-weight: 300;
+    color: #888;
+    width: 90%;
+    float: left;
+    padding-left: 20px;
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: Can we merge `padding-left` and `padding-right` by using `padding:`?

File: web/assets/css/bundles_list.css [hunk 25]
```diff
@@ -1,18 +1,153 @@
...
+    border: none;
+    float: left;
+    width: 10%;
+}
+
+.bundles-list-search button.bundles-list-search-button {
+    background: #f15a22;
+    color: #fff;
+    padding: 9px 45px 9px 45px;
+    border-top-left-radius: 0;
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: I think this can be improved. We can use `border-radius` instead.

File: web/assets/css/bundles_list.css [hunk 26]
```diff
@@ -1,18 +1,153 @@
...
+        width: 79%;
+    }
+
+    .bundles-list-search .input-group-addon {
+        width: 21%;
+        box-sizing: border-box;
+    }
+
+    .bundles-list-search button.bundles-list-search-button {
+        padding: 9px 0px 9px 0px;
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: Can be `padding: 9px 0;`

File: web/assets/css/bundles_list.css [hunk 27]
```diff
@@ -1,18 +1,153 @@
...
+    }
+
+    .bundles-list-container h2 {
+        font-size: 21px;
+        color: #fff;
+        background: #2a9dc7;
+        height: 40px;
+        line-height: 41px;
+        margin: 0;
+        padding: 0 17px 0 17px;
```
Comment (SylvainGuittard) [inline web/assets/css/bundles_list.css]: can be `padding: 0 17px;`

File: app/Resources/views/full/bundle_list.html.twig (hunk 10 above)
Comment (kmadejski) [inline app/Resources/views/full/bundle_list.html.twig]: Yes, you are right. It is my mistake, I will change displayed value to `Popularity`, in search conditions I will keep `stars` because it depends on Github stars.

Comment (damianz5) [inline app/Resources/views/full/bundle_list.html.twig]: depends on what we define as popular - most downloaded / most starred or combination of both. Using stars as popularity metric should be ok - it involves user action, and number of downloads is generated by composer

File: app/Resources/views/line/bundle_card.html.twig (hunk 11 above)
Comment (kmadejski) [inline app/Resources/views/line/bundle_card.html.twig]: It was done exactly the same as in design, but I agree, it will be better to remove spaces.

File: composer.json (hunk 14 above)
Comment (kmadejski) [inline composer.json]: Yes, I will fix it as soon as they will do the new release.

File: src/AppBundle/Service/Packagist/PackagistServiceProvider.php (hunk 19 above)
Comment (kmadejski) [inline src/AppBundle/Service/Packagist/PackagistServiceProvider.php]: Yes, we need timestamp but Packagist provides date and time in format: [ISO8601](http://php.net/manual/pl/class.datetime.php#datetime.constants.iso8601). I think the safest way is to create DateTime object from given date and time and then use it as timestamp using `format('U')`. See: [https://github.com/ezsystems/ezplatform-com/pull/101/files#diff-c9656b10c2a76d5eb42d06387e7ba085R87](https://github.com/ezsystems/ezplatform-com/pull/101/files#diff-c9656b10c2a76d5eb42d06387e7ba085R87)

File: src/AppBundle/Command/UpdateBundlesListCommand.php (hunk 15 above)
Comment (kmadejski) [inline src/AppBundle/Command/UpdateBundlesListCommand.php]: Sorry, but I don't see any other way. There is nothing like `$repository->sudo` in `$userService`. Obviously, we can create a user (with permissions to create/update content) and use this account instead of `admin`.

Comment (kmadejski) [inline src/AppBundle/Command/UpdateBundlesListCommand.php]: Yes, it is possible. I will refactor the whole Command. 
Thank you @alongosz, @webhdx, @Nattfarinn, @lserwatka.

File: app/Resources/views/full/bundle_list.html.twig (hunk 10 above)
Comment (SylvainGuittard) [inline app/Resources/views/full/bundle_list.html.twig]: So let's use `stars` for `Popularity` in the first iteration. We'll see later if we can improve.
